### PR TITLE
Add example to test normals and tangents (and fix WebGL)

### PR DESCRIPTION
**Pull request by erikdubbelboer**
This fixes normals and tangents in the WebGL2 shader. WebGPU is still broken, see: https://github.com/playcanvas/engine/issues/5735

Screenshot 2024-03-31 at 11 53 32

See also https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/NormalTangentTest

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: examples/src/examples/graphics/normals-and-tangents.mjs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    app.root.addChild(directionalLight);
+
+    app.start();
+    return app;
+}
+
+class NormalsAndTangentsExample {
+    static CATEGORY = 'Graphics';
+    static WEBGPU_ENABLED = true;
+    static DESCRIPTION = '<a href="https://github.com/KhronosGroup/glTF-Sample-Models/tree/main/2.0/NormalTangentTest">Model and textures by Ed Mackey</a>, copyright 2017-2018 Analytical Graphics Inc. CC-BY 4.0 <a href="https://creativecommons.org/licenses/by/4.0/">https://creativecommons.org/licenses/by/4.0/</a>.';
```
Comment (erikdubbelboer) [inline examples/src/examples/graphics/normals-and-tangents.mjs:117]: Depends on https://github.com/playcanvas/engine/pull/5922 to show the copyright in the description.

**Comment by erikdubbelboer** [comment]
I tried to find a way to fix these issues but I'm not sure where to start. Should this be fixed in the GLB parser or in somewhere else? If anyone can point me in the right direction I can try and fix the issues in this pull request as well.

**Comment by slimbuck** [comment]
Oh that's interesting. I never realised our lighting basis is wrong on the back faces. (We were aware of WebGPU issues though). Thanks for pointing this out!

**Comment by erikdubbelboer** [comment]
@slimbuck if you point me in the right direction, I can try and fix it. Are the normals and tangents calculated at a specific place? I couldn't easily find it as I'm not familiar enough with the source yet.

**Comment by slimbuck** [comment]
I would have expected backfaces to be handled by this bit of code here: https://github.com/playcanvas/engine/blob/main/src/scene/shader-lib/programs/lit-shader.js#L989

The actual tangents and binormals are calculated in this shader chunk:
https://github.com/playcanvas/engine/blob/main/src/scene/shader-lib/chunks/lit/frag/TBNderivative.js

**Comment by erikdubbelboer** [comment]
I have fixed the calculations for both WebGL and WebGPU. I have also fixed the WebGPU version looking so weird, this was caused by setting `mipmaps: true` on the `envAtlas`.

File: src/scene/shader-lib/chunks/lit/frag/TBNderivative.js [hunk 2]
```diff
@@ -20,6 +20,19 @@ void getTBN(vec3 tangent, vec3 binormal, vec3 normal) {
...
+        #else
+            dTBN = mat3(T * invmax, -B * invmax, gl_FrontFacing ? normal : -normal);
+        #endif
+    #else
+        #ifdef WEBGPU
+            dTBN = mat3(T * invmax, -B * invmax, -normal);
+        #else
+            dTBN = mat3(T * invmax, -B * invmax, normal);
+        #endif
+    #endif
```
Comment (slimbuck) [inline src/scene/shader-lib/chunks/lit/frag/TBNderivative.js:36]: Webgpu lighting space should not be different to webgl, so this normal negation is terribly confusing.

Are textures inverted (upside down) on webgpu vs webgl @mvaligursky? (If so then tangents and binormals as calculated would be inverted).

I think we need to get to the bottom of why normals needs flipping here and fix it at the source. My suspicion is we have an inverted transform somewhere else in the graphics pipeline under webgpu.

File: examples/src/examples/graphics/normals-and-tangents.mjs (hunk 1 above)
Comment (mvaligursky) [reply examples/src/examples/graphics/normals-and-tangents.mjs:117]: We don't need to display the attribution, we just need to have it stored in the file along with the glb. See this as an example of what we do: https://github.com/playcanvas/engine/blob/main/examples/assets/models/IridescentDishWithOlives.txt

File: src/scene/shader-lib/chunks/lit/frag/TBNderivative.js (hunk 2 above)
Comment (slimbuck) [reply src/scene/shader-lib/chunks/lit/frag/TBNderivative.js]: Actually I just noticed, looking in lit-shader.js, it seems we are already attempting to handle the normal for double-sided lighting case at https://github.com/playcanvas/engine/blob/main/src/scene/shader-lib/programs/lit-shader.js#L989. This calc must have a 🐛...

**Comment by mvaligursky** [comment]
I'd be definitely great to have this example as a test for normals / tangents.

File: src/scene/shader-lib/chunks/lit/frag/TBNderivative.js (hunk 2 above)
Comment (erikdubbelboer) [reply src/scene/shader-lib/chunks/lit/frag/TBNderivative.js]: You are right, it seems the issue is that it was handled in the wrong place. I have added a negatively scaled mesh to the example and I have gone though all variations of the shader to maker sure they all work correctly now.

File: src/scene/shader-lib/chunks/lit/frag/TBNfast.js [hunk 3]
```diff
@@ -1,5 +1,13 @@
 export default /* glsl */`
 void getTBN(vec3 tangent, vec3 binormal, vec3 normal) {
-    dTBN = mat3(tangent, binormal, normal);
+    #ifdef TWO_SIDED_LIGHTING
+        dTBN = mat3(
+            (gl_FrontFacing ? tangent  : -tangent ) * twoSidedLightingNegScaleFactor,
+            (gl_FrontFacing ? binormal : -binormal) * twoSidedLightingNegScaleFactor,
+            (gl_FrontFacing ? normal   : -normal  ) * twoSidedLightingNegScaleFactor
+        );
```
Comment (slimbuck) [inline src/scene/shader-lib/chunks/lit/frag/TBNfast.js:8]: Could simplify to something like
```suggestion
    dTBN = mat3(tangent, binormal, normal) * (-float(gl_FrontFacing) * twoSidedLightingNegScaleFactor);
```

Comment (willeastcott) [reply src/scene/shader-lib/chunks/lit/frag/TBNfast.js:8]: I'd vote for:

```
    dTBN = mat3(tangent, binormal, normal);

    #ifdef TWO_SIDED_LIGHTING
    dTBN *= ((gl_FrontFacing ? 1.0 : -1.0) * twoSidedLightingNegScaleFactor);
    #endif
```

Comment (erikdubbelboer) [reply src/scene/shader-lib/chunks/lit/frag/TBNfast.js:8]: Fixed.

**Comment by erikdubbelboer** [comment]
I have rebased the example on main with the new examples refactor. What needs to happen to get this merged now?

**Comment by willeastcott** [comment]
Thanks for doing that, @erikdubbelboer - we'll take another look ASAP! 😄

File: examples/src/examples/graphics/normals-and-tangents/config.mjs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+/**
+ * @type {import('../../../../types.mjs').ExampleConfig}
+ */
+export default {
+    WEBGPU_ENABLED: true,
```
Comment (mvaligursky) [inline examples/src/examples/graphics/normals-and-tangents/config.mjs:5]:
```suggestion
    HIDDEN: true,
    WEBGPU_ENABLED: true
```

Comment (mvaligursky) [reply examples/src/examples/graphics/normals-and-tangents/config.mjs:5]: I'm suggesting to hide it from public .. this is more of an internal example for us to validate how things work, and not an example to show the public how to use the engine.

**Comment by mvaligursky** [comment]
It seems the description only shows 'before' state, but not 'after' state - could you please update it.

**Comment by slimbuck** [comment]
Hi @erikdubbelboer,

Looking at the changes, I think it makes sense to handle twoSidedLighting in just one place - after dTBN has been calculated.

This is probably best done in a new shader chunk. So I went ahead and implemented this approach in https://github.com/playcanvas/engine/compare/main...slimbuck:erik-normals-and-tangents.

Could you take a look and let me know what you think? This approach also means we probably fix the case where `TBNObjectSpace.js` is used.

If you're happy with this I can push directly to your branch?

Thanks!

**Comment by erikdubbelboer** [comment]
@slimbuck great, I have merged your changes.

@mvaligursky I think this pull is ready now.

**Review by slimbuck — APPROVED** [review]

**Review by mvaligursky — APPROVED** [review]
This is much simpler now, thanks!

**Pull request merged**